src/components/document/annotation-layer.tsx

```diff
@@ -0,0 +1,170 @@
+import React, { useContext, useEffect, useState } from "react";
+
+import { ArrowAnnotationComponent } from "../annotations/arrow-annotation";
+import { TileApiInterfaceContext } from "../tiles/tile-api";
+import { useUIStore } from "../../hooks/use-stores";
+import { ArrowAnnotation, ArrowAnnotationType } from "../../models/annotations/arrow-annotation";
+import { ObjectBoundingBox, ClueObjectModel, ClueObjectType } from "../../models/annotations/clue-object";
+import { DocumentContentModelType } from "../../models/document/document-content";
+
+import "./annotation-layer.scss";
+
+interface IAnnotationButtonProps {
+  getObjectBoundingBox: (rowId: string, tileId: string, objectId: string) => ObjectBoundingBox | undefined;
+  key?: string;
+  objectId: string;
+  objectType?: string;
+  onClick?: (tileId: string, objectId: string, objectType?: string) => void;
+  rowId: string;
+  sourceObjectId?: string;
+  sourceTileId?: string;
+  tileId: string;
+}
+const AnnotationButton = observer(function AnnotationButton({
+  getObjectBoundingBox, objectId, objectType, onClick, rowId, sourceObjectId, sourceTileId, tileId
+}: IAnnotationButtonProps) {
+  const style = getObjectBoundingBox(rowId, tileId, objectId);
+  if (!style) return null;
+
+  const handleClick = () => onClick?.(tileId, objectId, objectType);
+
+  const source = sourceObjectId === objectId && sourceTileId === tileId;
+  const classes = classNames("annotation-button", { source });
+  return (
+    <button className={classes} onClick={handleClick} style={style} />
+  );
+});
+
+interface IAnnotationLayerProps {
+  content?: DocumentContentModelType;
+  documentScrollX?: number;
+  documentScrollY?: number;
+}
+export const AnnotationLayer = observer(function AnnotationLayer({
+  content, documentScrollX, documentScrollY
+}: IAnnotationLayerProps) {
+  const [_initialized, setInitialized] = useState(false);
+  useEffect(() => {
+    // Forces the annotation layer to rerender after initial load, getting access to the locations of elements.
+    setTimeout(() => setInitialized(true));
+  }, []);
+  const [sourceTileId, setSourceTileId] = useState("");
+  const [sourceObjectId, setSourceObjectId] = useState("");
+  const [sourceObjectType, setSourceObjectType] = useState<string | undefined>();
+  const ui = useUIStore();
+  const tileApiInterface = useContext(TileApiInterfaceContext);
+
+  const rowIds = content?.rowOrder || [];
+
+  function getObjectBoundingBox(
+    rowId: string, tileId: string, objectId: string
+  ) {
+    const rowSelector = `[data-row-id='${rowId}']`;
+    const rowElements = document.querySelectorAll(rowSelector);
+    if (rowElements.length !== 1) return undefined;
+    const rowElement = (rowElements[0] as HTMLElement);
+  
+    const tileSelector = `[data-tool-id='${tileId}']`;
+    const tileElements = document.querySelectorAll(tileSelector);
+    if (tileElements.length !== 1) return undefined;
+    const tileElement = (tileElements[0] as HTMLElement);
```

Comment:
This is where things are failing in a curriculum with two panels. The issue is that both documents have elements with the same `data-row-id` and `data-tool-id`, and this code bails when it can't find a unique element it's looking for.

Reply:
I was able to solve this issue by giving the document content component a class based on whether it's read only or not. I then check for that class in these `querySelectorAll`s.

I don't think this solution is particular elegant, though. It's also a bit frail, because it would fail if there were ever more than one read only copy of a document rendered at the same time. Does this ever happen?

Anyway, I welcome alternative suggestions on how to solve this issue. There's a separate story for handling read only documents, so it might make sense to push a better solution to that story if it would be time consuming to implement.